### ISSUE TITLE
Use mrustc to bootstrap Rust compiler

### ISSUE THREAD

**Issue by tadfisher** [NixOS/nixpkgs#72606]
[mrustc](https://github.com/thepowersgang/mrustc) is a minimal Rust compiler that can compile rustc and cargo 1.19 and 1.29, which can replace our binary bootstrap compiler and give a trusted root to the bootstrap chain.

**Comment by ivan** [comment]
https://github.com/guix-mirror/guix/blob/7dadce383fc5f00b795c8fa1209033469731c14b/gnu/packages/rust.scm has a many-phase approach to building a recent rustc starting with mrustc

**Comment by ProgVal** [comment]
still active (see #85542)

**Comment by poelzi** [comment]
got a fix :) CPU says BRRRRRRR

**Comment by whispersofthedawn** [comment]
just as an update on what we've gathered from various discussions over time:
- this is easily technically *feasible*. there have been many proofs of concept over the years, and we have plenty of work to base on from guix, stagex, etc. many people have implemented this in nix as well.
- this is not *maintainable*: in particular, as a predominantly single-person project, mrustc often gets far behind rustc in versions. before the release of mrustc 0.12 adding support for rustc 1.90, the current rust version was 1.93, which means that the bootstrap chain from mrustc → latest stable would have required 20 rust versions (1.74..1.93) and five llvm versions (17..21), some of which had already been removed from nixpkgs. we try not to keep llvm versions around for too long, and have enough trouble removing them (and keeping them working as the rest of nixpkgs updates, etc.) as it is. we do not want to make this situation worse. thus, until mrustc has a history of keeping up better, we have no reason to believe that this wouldn't be maintenance hell.
- (personally) this would be very annoying, as currently rustc takes 40-60 minutes to build on my machine. this is something i do regularly for various purposes on `staging`. if we add ten rustc versions (close to where we are now!), that trivially becomes at least a five hour chain on our machine (we've tested this before), and nearly ten hours in the pre-mrustc-0.12 state. it'd be slightly less bad on hydra since it has much better builders, but selfishly we'd rather not lock ourselves out from being able to do this testing on a regular basis.

tl;dr we'd really like to see source bootstraps of rust as well, but it doesn't seem particularly maintainable and reasonable in the current state of things.

**Comment by Atemu** [comment]
> * five llvm versions (17..21), some of which had already been removed from nixpkgs.

A thought I had related to this is that this isn't necessarily a huge issue as these would exclusively be used for bootstrap and not otherwise exposed; just as we also use ancient gcc versions for our C compiler bootstrap. These would also only need to be able to compile rust and nothing else which I imagine allows for quite some trimming down.

> * currently rustc takes 40-60 minutes to build on my machine. this is something i do regularly for various purposes on `staging`. if we add ten rustc versions (close to where we are now!), that trivially becomes at least a five hour chain on our machine (we've tested this before)

May I ask which CPU this is? I'm asking because we need some baseline to compare when discussing perf.

I'd say a 6 y/o mid-range chip like the 5800X would be a decent baseline for the minimum that we can expect a contributor to use when regularly doing such builds.

Has this rustc chain been optimised for build time? Only the last stage would need to be tuned to generate fast binaries. The slop PR author claimed to have had 7min per intermediate rustc on a quite fast CPU. If it took twice that on a less performant CPU, that wouldn't be nearly as bad?

**Comment by Atemu** [comment]
FWIW my 3 y/o 7950x3D can build the regular top-level `rustc.passthru.unwrapped` with binary speed tuning, docs and everything in ~11min. That's still quite a fast CPU but "only" 2-3× faster than a 5800X at compilation. I would not be surprised if a 5800X could finish a trimmed build in a similar amount of time.

**Comment by pyrox0** [comment]
Well, that's using a binary rustc that is optimized(the one downloaded from rust-lang.org), so I imagine that building rustc with a non-optimized previous rustc may change the results(not that I know how much the impact would be, it's still something to keep in mind)
